### PR TITLE
Create topics always MODINVSTOR-839

### PR DESCRIPTION
**Pull request by adamdickmeiss**
Create topics always and ignore if topic already exists, thus having
the effect of "create if not exists".

### CODE REVIEW

**Comment by adamdickmeiss** [comment]
I think I've seen this test failure before.. A known issue or cause of this PR?
```
Failed tests: 

  ItemEffectiveCallNumberComponentsTest.canCalculateEffectiveCallNumberPropertyOnUpdate:261 

Expected: is <{"id":"6f096389-2103-46d6-b1ca-18fc1f93dd49","_version":2,"hrid":"it00000000034","holdingsRecordId":"b2f8cdc0-465d-4f9b-9937-a644910e8a06","formerIds":[],"effectiveShelvingOrder":"","itemLevelCallNumber":"initItCN","effectiveCallNumberComponents":{"callNumber":"initItCN"},"yearCaption":[],"administrativeNotes":[],"notes":[],"circulationNotes":[],"status":{"name":"Available","date":"2021-11-25T15:08:47.885+00:00"},"materialTypeId":"9d33cc06-954e-4848-817c-cec3170c078f","permanentLoanTypeId":"19cf1d4d-aaa0-42bd-aea7-744ef62fcf7e","temporaryLocationId":"f7a0bd1c-d614-47f8-9a81-324882bcadd3","effectiveLocationId":"f7a0bd1c-d614-47f8-9a81-324882bcadd3","electronicAccess":[],"statisticalCodeIds":[],"metadata":{"createdDate":"2021-11-25T15:08:47.885+00:00","updatedDate":"2021-11-25T15:08:47.885+00:00"},"instanceId":"f1953254-c26d-44bc-8ea3-92cd42962417"}>

     but: was <{"instanceId":"f1953254-c26d-44bc-8ea3-92cd42962417","id":"6f096389-2103-46d6-b1ca-18fc1f93dd49","_version":1,"hrid":"it00000000034","holdingsRecordId":"b2f8cdc0-465d-4f9b-9937-a644910e8a06","formerIds":[],"effectiveShelvingOrder":"","itemLevelCallNumber":"initItCN","effectiveCallNumberComponents":{"callNumber":"initItCN"},"yearCaption":[],"administrativeNotes":[],"notes":[],"circulationNotes":[],"status":{"name":"Available","date":"2021-11-25T15:08:47.885+00:00"},"materialTypeId":"9d33cc06-954e-4848-817c-cec3170c078f","permanentLoanTypeId":"19cf1d4d-aaa0-42bd-aea7-744ef62fcf7e","temporaryLocationId":"f7a0bd1c-d614-47f8-9a81-324882bcadd3","effectiveLocationId":"f7a0bd1c-d614-47f8-9a81-324882bcadd3","electronicAccess":[],"statisticalCodeIds":[],"metadata":{"createdDate":"2021-11-25T15:08:47.885+00:00","updatedDate":"2021-11-25T15:08:47.885+00:00"}}>
```

File: src/test/java/org/folio/services/kafka/topic/KafkaAdminClientServiceTest.java [hunk 1]
```diff
@@ -43,7 +43,7 @@ public void shouldNotCreateTopicIfAlreadyExist(TestContext testContext) {
     createKafkaTopicsAsync(mockClient)
       .onFailure(testContext::fail)
       .onComplete(testContext.asyncAssertSuccess(notUsed -> {
-        verify(mockClient, times(0)).createTopics(anyList());
+        verify(mockClient, times(1)).createTopics(anyList());
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/services/kafka/topic/KafkaAdminClientServiceTest.java:46]: Given that the topics are now _meant to be created_ I think the name should be changed to reflect that intent

File: src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java [hunk 2]
```diff
@@ -78,24 +68,19 @@ public KafkaAdminClientService(Supplier<KafkaAdminClient> clientFactory) {
...
-        return succeededFuture();
-      }
-
-      log.info("Creating topics {}", topicsToCreate);
-      return kafkaAdminClient.createTopics(topicsToCreate);
-    });
+    return kafkaAdminClient.createTopics(topics)
+      .recover(x -> {
+        if (x instanceof org.apache.kafka.common.errors.TopicExistsException) {
+          return Future.succeededFuture();
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java:79]: Could we log that this error has occurred?

As I understand it, this change introduced the chance that the creation of one topic fails, leaving other topics not created and yet the Tenant API request will succeed stating that the module should work for the tenant.

Comment (julianladisch) [reply src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java:79]: Kafka waits until all topics have been processed before throwing a TopicExistsException. Only if no other topic caused some other exception a TopicExistsException is thrown.
https://github.com/apache/kafka/blob/3.0.0/trogdor/src/main/java/org/apache/kafka/trogdor/common/WorkerUtils.java#L215

**Comment by julianladisch** [comment]
> I have a loop implementation now, but it doesn't work as there are cases where we did not previously purge and, hence, just want to leave the topics as is (only create if not exists).

> https://github.com/folio-org/mod-inventory-storage/tree/recreate-topics-until-success

The loop need to query listTopics().

See the loop that the Kafka people use: https://github.com/apache/kafka/blob/3.0.0/streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java#L441-L449

File: src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java (hunk 2 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java:79]: @julianladisch Thank you for the explanation and link

Comment (adamdickmeiss) [reply src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java:79]: Thx @julianladisch . I made a test for this, although not a bullet proof one. I'll add a log msg message. level? log? warn?

File: src/test/java/org/folio/services/kafka/topic/KafkaAdminClientServiceTest.java (hunk 1 above)
Comment (adamdickmeiss) [reply src/test/java/org/folio/services/kafka/topic/KafkaAdminClientServiceTest.java:46]: yep

**Comment by adamdickmeiss** [comment]
> > I have a loop implementation now, but it doesn't work as there are cases where we did not previously purge and, hence, just want to leave the topics as is (only create if not exists).
> 
> > https://github.com/folio-org/mod-inventory-storage/tree/recreate-topics-until-success
> 
> The loop need to query listTopics().
> 
> See the loop that the Kafka people use: https://github.com/apache/kafka/blob/3.0.0/streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopicManager.java#L441-L449

The first code you showed also had a deleteTopics in each loop.. But we can't use this strategy as the Kafka example assumes topics previously removed.. Which is the unusual case for his (upgrade case).

File: src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java (hunk 2 above)
Comment (julianladisch) [reply src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java:79]: info.

**Comment by julianladisch** [comment]
> The first code you showed also had a deleteTopics in each loop.. But we can't use this strategy as the Kafka example assumes topics previously removed.. Which is the unusual case for this (upgrade module case).

Our old code skips topics that show up in listTopics(): https://github.com/folio-org/mod-inventory-storage/blob/0ba4a59045c38b1ff2061d999d70f6d975aa46d0/src/main/java/org/folio/services/kafka/topic/KafkaAdminClientService.java#L87

**Comment by julianladisch** [comment]
This should be sufficient, no need to call deleteTopic.

File: src/test/java/org/folio/rest/api/KafkaTopicsExistsTest.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      if (cause instanceof org.apache.kafka.common.errors.TopicExistsException) {
+        return null;
+      }
+      throw new RuntimeException(cause);
+    });
+    // test that T3 got inserted, while T1 already being present
+    future.compose(x -> kafkaAdminClient.listTopics())
+      .onComplete(context.asyncAssertSuccess(topics ->
+        context.assertTrue(topics.containsAll(List.of("T1", "T2", "T3")))
+      )).mapEmpty();
```
Comment (julianladisch) [inline src/test/java/org/folio/rest/api/KafkaTopicsExistsTest.java:62]: `future` variable and `.mapEmpty()` are not needed.

File: src/test/java/org/folio/rest/api/KafkaTopicsExistsTest.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package org.folio.rest.api;
```
Comment (julianladisch) [inline src/test/java/org/folio/rest/api/KafkaTopicsExistsTest.java:1]: This doesn't look like an API test for me and could go into package org.folio.services.kafka.topic.

Comment (adamdickmeiss) [reply src/test/java/org/folio/rest/api/KafkaTopicsExistsTest.java:1]: I put it there as it uses StorageTestSuite / TestBase.. You still think it should go there?

File: src/test/java/org/folio/rest/api/KafkaTopicsExistsTest.java (hunk 3 above)
Comment (adamdickmeiss) [reply src/test/java/org/folio/rest/api/KafkaTopicsExistsTest.java:62]: ok

File: src/test/java/org/folio/rest/api/KafkaTopicsExistsTest.java (hunk 4 above)
Comment (adamdickmeiss) [reply src/test/java/org/folio/rest/api/KafkaTopicsExistsTest.java:1]: Moving it as everything is sufficiently public.

File: src/test/java/org/folio/rest/api/StorageTestSuite.java [hunk 5]
```diff
@@ -88,7 +89,8 @@
   LegacyItemEffectiveLocationMigrationScriptTest.class,
   IterationJobRunnerTest.class,
   AuthorityStorageTest.class,
-  SampleDataTest.class
+  SampleDataTest.class,
+  KafkaTopicsExistsTest.class
```
Comment (julianladisch) [inline src/test/java/org/folio/rest/api/StorageTestSuite.java:93]: No need to put KafkaTopicsExistsTest into the StorageTestSuite.
mvn automatically runs all org/folio/services tests: https://github.com/folio-org/mod-inventory-storage/blob/v22.0.2/pom.xml#L458

File: src/test/java/org/folio/services/kafka/topic/KafkaTopicsExistsTest.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        ))
+      ).otherwise(cause -> {
+        if (cause instanceof org.apache.kafka.common.errors.TopicExistsException) {
+          return null;
+        }
+        throw new RuntimeException(cause);
+      }).compose(x -> kafkaAdminClient.listTopics())
+      .onComplete(context.asyncAssertSuccess(topics ->
+        context.assertTrue(topics.containsAll(List.of("T1", "T2", "T3")))
+      )).mapEmpty();
```
Comment (julianladisch) [inline src/test/java/org/folio/services/kafka/topic/KafkaTopicsExistsTest.java:61]: `.mapEmpty()` is not needed.

File: src/test/java/org/folio/rest/api/StorageTestSuite.java (hunk 5 above)
Comment (adamdickmeiss) [reply src/test/java/org/folio/rest/api/StorageTestSuite.java:93]: ok

**Review by julianladisch — APPROVED** [review]
Looks good, thanks!

**Pull request merged**